### ISSUE TITLE
Implement a heartbeat mechanism to keep idle connections alive

### ISSUE THREAD

**Issue by danielxvu** [peer-calls/peer-calls#262]
**Describe the bug**

For Kubernetes clusters hosted on Google Kubernetes Engine, hereby referred to as GKE, the backend service timeout is [30 seconds by default](https://cloud.google.com/load-balancing/docs/backend-service#timeout-setting). As a result, any WebSocket connection that idles will be automatically closed after 30 seconds, forcing a reconnect. There are two ways to resolve this:

1. As a workaround, [increase the backend service timeout](https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-features#timeout).
2. Implement an application-level heartbeat mechanism. The front-end is already sending a ping periodically, so a PR to address this could just update the server accordingly. But it's probably better for the server to send the ping so that it can clean up stale connections when clients fail to send a pong back after a certain amount of pings and/or time.

This is encountered commonly enough for WebSocket apps deployed to Kubernetes clusters that [IBM has documented the issue here](https://cloud.ibm.com/docs/containers?topic=containers-cs_ingress_websocket). I can confirm myself that I ran into this issue for an unrelated project using a WebSocket server deployed to AWS' managed Kubernetes service and that a heartbeat mechanism will fix this issue for sure. Happy to open a PR if you're willing to accept it.

See the following output for an example of what's happening:

```console
│ 2022-04-20T04:53:54.976301Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:53:54.977463Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:53:55.020683Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:54:24.979744Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:54:24.979798Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:54:59.116038Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:54:59.118349Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:54:59.165173Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:55:29.120557Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:55:29.120597Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:55:31.192008Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:55:31.193457Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:55:31.236369Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:56:01.196246Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:56:01.196298Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:56:35.329223Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:56:35.331623Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:56:35.387500Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:57:05.331969Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:57:05.332018Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:57:07.395933Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:57:07.397254Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:57:07.448145Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:57:37.398760Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:57:37.398831Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:58:11.534466Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:58:11.536506Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:58:11.589494Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:58:41.537675Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:58:41.537715Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:58:43.600252Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:58:43.601801Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:58:43.643807Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:59:13.602861Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:59:13.602899Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T04:59:15.675848Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T04:59:15.677216Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T04:59:15.728389Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T04:59:45.679293Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T04:59:45.679330Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T05:00:19.832378Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T05:00:19.834768Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T05:00:19.886100Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T05:00:49.835859Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T05:00:49.835904Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T05:03:00.212018Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T05:03:00.214576Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T05:03:00.256336Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T05:03:30.214772Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T05:03:30.214846Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T05:03:32.278332Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T05:03:32.279668Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T05:03:32.332569Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T05:04:02.282700Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T05:04:02.282745Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST                                                                                                                                                                                                                       │
│ 2022-04-20T05:04:04.350007Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Enter room_id=FAST                                                                                                                                                                                                                      │
│ 2022-04-20T05:04:04.351252Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] New websocket connection room_id=FAST                                                                                                                                                                                                   │
│ 2022-04-20T05:04:04.396989Z  info [bsocket_handler:mesh] [RFWmhiHJ0teOp8uuEkUTv7] Got clients: map[RFWmhiHJ0teOp8uuEkUTv7:Daniel Vu] room_id=FAST                                                                                                                                                                         │
│ 2022-04-20T05:04:34.353063Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Remove room_id=FAST                                                                                                                                                                                                                     │
│ 2022-04-20T05:04:34.353103Z  info [        main:mux:wss] [RFWmhiHJ0teOp8uuEkUTv7] Exit room_id=FAST  
```

**Bug Report Checklist**

n/a

**Steps To Reproduce**

Deploy a Kubernetes cluster to GKE using the example manifests in this repo. After doing so, join a room as the sole participant and leave the browser tab open for a few minutes.

**Expected behavior**

For a WebSocket connection to a GKE-hosted peer-calls pod, said connection should not be closed automatically after 30 seconds if idle.

**Screenshots**

![image](https://user-images.githubusercontent.com/56174945/164155150-b3439e38-7b08-44c0-b73c-0279f044d0cc.png)


**Additional context**

Helpful links explaining the need for a heartbeat mechanism:

- https://stackoverflow.com/a/46112000
- https://stackoverflow.com/a/35826016
- https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets

**Comment by prologic** [comment]
I _think_ implementing 2) is probably the right direction here, and given I've had a lot of experience myself with Websockets and Go of late I _might_ be able to help with this... 🤔

**Comment by jeremija** [comment]
@danielxvu Thanks for reporting this. (2) makes sense to me as well. Can you take a look at #267 and see if that works for you?

**Comment by danielxvu** [comment]
@jeremija Will do, thanks for getting around to this!

**Comment by danielxvu** [comment]
I'm testing this today, will report back shortly with the results.

**Comment by danielxvu** [comment]
@jeremija Unfortunately something is still causing the WebSocket connection to be closed periodically. I do see the ping-pong happening as you can see from the following screenshot:

![image](https://user-images.githubusercontent.com/56174945/167968913-758452cf-2a80-4482-a684-1cefbc9eb7aa.png)

But as you can see from the following server-side logs, the connection keeps re-initializing.

```
2022-05-12T00:30:30.527448Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:30.528149Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:30.528614Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:30.557304Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:30.558400Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:30.559309Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:30.605086Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:30.625493Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:30.688417Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:30.745391Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:30.784003Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:44.372392Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Remove room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:30:44.372424Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Exit room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:30:46.459361Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Enter room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:30:46.459416Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] New websocket connection room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:30:46.500015Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Got clients: map[8fq82rV4RhLLIJ5uPOjfO1:Daniel Vu LiX0ngMVYphU2AzbgL1G11:wally] room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:30:46.584771Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:46.585402Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:46.585860Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:46.616279Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:46.617586Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:46.618042Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:46.662632Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:46.683157Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:46.741617Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:30:46.800747Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:30:46.839358Z  info [bsocket_handler:mesh] [LiX0ngMVYphU2AzbgL1G11] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=8fq82rV4RhLLIJ5uPOjfO1
2022-05-12T00:31:00.430260Z  info [        main:mux:wss] [LiX0ngMVYphU2AzbgL1G11] Remove room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:00.430306Z  info [        main:mux:wss] [LiX0ngMVYphU2AzbgL1G11] Exit room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:02.653013Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:02.653716Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:02.654270Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:02.707912Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:02.788184Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:02.883213Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:16.461173Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Remove room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:16.461213Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Exit room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:18.525423Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Enter room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:18.526747Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] New websocket connection room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:18.572961Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Got clients: map[8fq82rV4RhLLIJ5uPOjfO1:Daniel Vu LiX0ngMVYphU2AzbgL1G11:wally] room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:18.644728Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:18.645771Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:18.647281Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:18.738470Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:18.845429Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.715724Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.716613Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.717217Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.765578Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.830200Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:34.929329Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:48.527380Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Remove room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:48.527437Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Exit room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:50.605236Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] Enter room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:50.607233Z  info [        main:mux:wss] [8fq82rV4RhLLIJ5uPOjfO1] New websocket connection room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:50.649907Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Got clients: map[8fq82rV4RhLLIJ5uPOjfO1:Daniel Vu LiX0ngMVYphU2AzbgL1G11:wally] room_id=jPrCuNJL5cPEHYE9HDiUV6
2022-05-12T00:31:50.727108Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:50.728104Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:50.728582Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:50.824101Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
2022-05-12T00:31:50.928306Z  info [bsocket_handler:mesh] [8fq82rV4RhLLIJ5uPOjfO1] Send signal to room_id=jPrCuNJL5cPEHYE9HDiUV6 target_client_id=LiX0ngMVYphU2AzbgL1G11
```

Here are the client-side logs:

```
InsertableStreams worker initialized https://peercalls.example.net/call/jPrCuNJL5cPEHYE9HDiUV6
audio.worklet.js:1 VUMeter initiated
index.js?v4.1.2-1-g7833718-dirty:2 play rejected AbortError
jPrCuNJL5cPEHYE9HDiUV6:1 Uncaught (in promise) DOMException: The play() request was interrupted by a call to pause(). https://goo.gl/LdLk22
```

**Comment by danielxvu** [comment]
@jeremija I'm wondering if GKE wants us to set a value no matter what for the [backend service timeout](https://cloud.google.com/load-balancing/docs/https#timeouts_and_retries) for whenever an HTTP connection gets upgraded to a WebSocket. The docs say:

> If you are using WebSockets, the backend service timeout should be set to the maximum duration of a WebSocket, idle or active.

I tested increasing the timeout on master and doing so did resolve the issue. But I've deployed GKE LoadBalancer services for other WebSocket app deployments before and didn't have to increase the backend service timeout of 30 seconds on such occasions. So I'm wondering if the timeout increase is just a workaround and not actually addressing an underlying issue.

That being said, I think #267 could still be needed for Amazon EKS or IBM clusters and thus worth having regardless of whether it fixes this specific GKE issue.
